### PR TITLE
Add sectioned scoped timer

### PR DESCRIPTION
**Pull request by LGM-Doyle**
This PR does three things: 
- makes the time threshold for display of `ScopedTimer` a parameter of the constructor
- adds `SectionedScopedTimer` as a `ScopedTimer` with separately timed sub-sections.
- uses `SectionedScopedTimer` in `MapWnd::InitTurn()` as an example.

`SectionedScopedTimer` is a `ScopedTimer` with an additional member function`EnterSection(<section name>)`, which creates/re-enters a separately timed sub section, with an arbitrary name.

When `SectionedScopedTimer` goes out of scope and is destroyed it prints a table of times.  The table includes its total time and the time of any subsections that exceed the threshold parameter.

An example table looks like this:

```
         Title - Section  time: xxxx ms
         Title - Section2 time: xxxx ms
         Title            time: xxxx ms
```

SectionedScopedTimer was the profiling tool that I used while working on the split fleet timing PR #863.  I stripped it out of that PR to submit here separately.

### CODE REVIEW

File: util/ScopedTimer.cpp [hunk 1]
```diff
@@ -3,28 +3,223 @@
...
+    }
+
+    bool ShouldOutput(const boost::chrono::nanoseconds & duration) {
+        return (duration >= m_threshold && (m_enable_output || GetOptionsDB().Get<bool>("verbose-logging")));
     }
-    boost::timer    m_timer;
-    std::string     m_name;
-    bool            m_always_output;
+
+    void FormatDuration(std::stringstream &ss, const boost::chrono::nanoseconds & duration) {
```
Comment (geoffthemedio) [inline util/ScopedTimer.cpp]: &ss -> & ss

File: util/ScopedTimer.cpp [hunk 2]
```diff
@@ -3,28 +3,223 @@
...
+        SectionTable::iterator m_curr;
+
+        // Names of the sections in order or creation.
+        std::vector<std::string> m_section_names;
+
+    };
+
+public:
+    SectionedScopedTimerImpl(const std::string& timed_name, bool enable_output,
+                    boost::chrono::microseconds threshold) :
```
Comment (geoffthemedio) [inline util/ScopedTimer.cpp]: indenting to (

File: util/ScopedTimer.cpp [hunk 3]
```diff
@@ -3,28 +3,223 @@
...
+
+
+
+class SectionedScopedTimer::SectionedScopedTimerImpl : public ScopedTimer::ScopedTimerImpl {
+
+    /** Sections store a time and a duration for each section of the elapsed time report.*/
+    struct Sections {
+        Sections(const boost::chrono::high_resolution_clock::time_point &now,
+                 const boost::chrono::nanoseconds& time_from_start) :
+            m_table(), m_section_start(now), m_curr(), m_section_names() {
```
Comment (geoffthemedio) [inline util/ScopedTimer.cpp]: { to next line

File: util/ScopedTimer.cpp [hunk 4]
```diff
@@ -3,28 +3,223 @@
...
+
+            // Create a dummy "" section so that m_curr is always a valid iterator.
+            std::pair<SectionTable::iterator, bool> curr = m_table.insert(std::make_pair("", time_from_start));
+            m_curr = curr.first;
+        }
+
+        /** Add time to the current section and then setup the new section. */
+        void Accumulate(const boost::chrono::high_resolution_clock::time_point &now,
+                        const std::string & section_name) {
```
Comment (geoffthemedio) [inline util/ScopedTimer.cpp]: ...{

File: util/ScopedTimer.cpp [hunk 5]
```diff
@@ -3,28 +3,223 @@
...
+
+        //Stop the final section.
+        EnterSection("");
+
+        //Print the section times followed by the total time elapsed.
+
+        // Find the longest name to right align the times.
+        size_t longest_section_name(0);
+        for (std::vector<std::string>::const_iterator it = m_sections->m_section_names.begin();
+             it != m_sections->m_section_names.end(); ++it) {
```
Comment (geoffthemedio) [inline util/ScopedTimer.cpp]: ...

File: util/ScopedTimer.cpp [hunk 6]
```diff
@@ -3,28 +3,223 @@
...
+            longest_section_name = std::max(longest_section_name, it->size());
+        }
+
+        for (std::vector<std::string>::const_iterator it = m_sections->m_section_names.begin();
+             it != m_sections->m_section_names.end(); ++it) {
```
Comment (geoffthemedio) [inline util/ScopedTimer.cpp]: ...

**Comment by LGM-Doyle** [comment]
I've made the changes.

**Pull request merged**